text/0045-apply-keyword.md

```diff
@@ -0,0 +1,134 @@
+# Add Apply keyword to Cedar language
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): 
+
+## Timeline
+
+- Started: 2024-01-25
+
+## Summary
+
+Add a keyword that allows denoted policy statements to have no impact on the
+authorization decision, yet be logged in the diagnostics for later annotation
+aggregation.
+
+## Basic example
+
+```cedar
+@annotation("value")
+apply (
+    principal,
+    action,
+    resource
+);
+```
+
+## Motivation
+
+I want to be able to create policy statements that will add annotations while
+not affecting the authorization outcome.  Here are some examples:
+
+1. I want to be able to require MFA anytime someone accesses a certain table.
+However, I want to be able to define permission for this table elsewhere.
+
+2. I want to be able to send out an email everytime something is denied.
+However, I want to be able to define what things are forbidden elsewhere.
+
+3. I want to be able to send out an email regardless of if a specific event
+is allowed or denied.
+
+## Detailed design
+
+There are three cases that are added to the language syntax:
+
+1. Apply on permit
+
+```cedar
+@mfa("true")
+apply on permit (
+    principal,
+    action == Postgres::Action::"select",
+    resource == Postgres::Table::"example.com:5432/db"
+);
+```
```

Comment:
For this example in particular, I'm not convinced this is a good motivating example.  The goal stated above was "I want to ... require MFA anytime someone accesses a certain table."  I think this should instead be done with something like `when { context.mfa == true }` on the `permit` policy(ies) in question; or if that is undesirable, then by adding a `forbid` policy along the lines of 
```
forbid ( 
    principal, 
    action == Postgres::Action::"select", 
    resource == Postgres::Table::"example.com:5432/db" 
) when { 
    context.mfa == false 
};
```

This expresses the authorization-relevant logic directly in the Cedar policy(ies), rather than requiring the application to correctly interpret policy annotations.  In general it is a best-practice to have all of the logic relevant to an authorization decision expressed in Cedar itself, not expressed in application logic surrounding the call to Cedar.

Reply:
Yeah, if the user was previously authenticated with MFA, I'd be able to include the MFA information in the context.

The case I'm describing is when you want to trigger an MFA after the cedar is evaluated, so I can have the provider pop up a window on your phone with the action "allow this query".  The result of the MFA then modifies the final decision.

Reply:
Cedar partial evaluation is designed for this use-case.  We can leave `context.mfa` unknown, and if the overall authentication decision ends up depending on `context.mfa`, partial evaluation will report an unknown with a residual like `context.mfa == true`.  Your application could take this unknown, realize it needs the user to authenticate with MFA, and then when appropriate retry the `is_authorized` call.

You could also do this without partial evaluation.  You could add the `forbid` policy I wrote above, but annotate it with something that tells you that if this `forbid` actually controls the decision, you should ask the user to authenticate with MFA.  Once they do, you can retry the `is_authorized` call.

Reply:
In the first case, when you say "your application could take this unknown" is that meaning, by reviewing the AST we'd be able to know what fields are unknown?  I hadn't read about the partial evaluation before, so that is interesting.  It seems like that could work, though I'd also want some annotation somewhere to indicate what the MFA prompt should be.  I'm imagining:

```
@mfa("query a table")
permit (
    principal, 
    action == Postgres::Action::"select", 
    resource == Postgres::Table::"example.com:5432/db" 
) when { 
    context.mfa == true 
};
```

In the second case are you saying the cedar would look like:

```
@mfa("query a table")
forbid ( 
    principal, 
    action == Postgres::Action::"select", 
    resource == Postgres::Table::"example.com:5432/db" 
) when { 
    context.mfa == false 
};

permit (
    principal, 
    action == Postgres::Action::"select", 
    resource == Postgres::Table::"example.com:5432/db" 
) when { 
    context.mfa == true 
};
```

So the forbid would give me the hint to gather the MFA, then to re-run the authorization afterwards.

If I'm capturing the concepts correctly, I can add them to the alternatives.

Reply:
Yes, that's right.  In the second case the `permit` doesn't even need the `when` clause, because the `forbid` overrides it.  (The `when` clause doesn't hurt, of course.)

Reply:
I've added both of these alternatives to the RFC.  Thanks!